core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala

```diff
@@ -0,0 +1,279 @@
+    controllerContext.partitionReplicaAssignment.put(partition, Seq(brokerId))
+    partitionState.put(partition, OfflinePartition)
+    val leaderAndIsr = LeaderAndIsr(LeaderAndIsr.NoLeader, List(brokerId))
+    val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controllerEpoch)
+    controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+
+    EasyMock.expect(mockControllerBrokerRequestBatch.newBatch())
+    EasyMock.expect(mockZkUtils.getTopicPartitionStates(partitions))
+      .andThrow(new ZookeeperClientException(""))
+
+    EasyMock.expect(mockControllerBrokerRequestBatch.sendRequestsToBrokers(controllerEpoch))
+    EasyMock.replay(mockZkUtils, mockControllerBrokerRequestBatch)
+
+    partitionStateMachine.handleStateChanges(partitions, OnlinePartition, Option(OfflinePartitionLeaderElectionStrategy))
+    EasyMock.verify(mockZkUtils, mockControllerBrokerRequestBatch)
+    assertEquals(OfflinePartition, partitionState(partition))
+  }
+
+  @Test
+  def testOfflinePartitionToOnlinePartitionTransitionErrorCodeFromStateLookup(): Unit = {
+    controllerContext.liveBrokers = Set(TestUtils.createBroker(brokerId, "host", 0))
+    controllerContext.partitionReplicaAssignment.put(partition, Seq(brokerId))
+    partitionState.put(partition, OfflinePartition)
+    val leaderAndIsr = LeaderAndIsr(LeaderAndIsr.NoLeader, List(brokerId))
+    val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controllerEpoch)
+    controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+
+    val stat = new Stat(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
+    EasyMock.expect(mockControllerBrokerRequestBatch.newBatch())
+    EasyMock.expect(mockZkUtils.getTopicPartitionStates(partitions))
+      .andReturn(Seq(GetDataResponse(Code.NONODE.intValue(), null, partition, TopicPartitionStateZNode.encode(leaderIsrAndControllerEpoch), stat)))
+
+    EasyMock.expect(mockControllerBrokerRequestBatch.sendRequestsToBrokers(controllerEpoch))
+    EasyMock.replay(mockZkUtils, mockControllerBrokerRequestBatch)
+
+    partitionStateMachine.handleStateChanges(partitions, OnlinePartition, Option(OfflinePartitionLeaderElectionStrategy))
+    EasyMock.verify(mockZkUtils, mockControllerBrokerRequestBatch)
+    assertEquals(OfflinePartition, partitionState(partition))
+  }
+
+  @Test
+  def testOfflinePartitionToNonexistentPartitionTransition(): Unit = {
+    partitionState.put(partition, OfflinePartition)
+    partitionStateMachine.handleStateChanges(partitions, NonExistentPartition)
+    assertEquals(NonExistentPartition, partitionState(partition))
+  }
+
+  @Test
+  def testInvalidOfflinePartitionToNewPartitionTransition(): Unit = {
+    partitionState.put(partition, OfflinePartition)
+    partitionStateMachine.handleStateChanges(partitions, NewPartition)
+    assertEquals(OfflinePartition, partitionState(partition))
+  }
+
```

Comment:
It will be useful to add a test for the controlled shutdown case, which involves setting controllerContext.shuttingDownBrokerIds and making the state transition to online.

Reply:
Done.